crates/core/tedge_agent/src/file_transfer_server/http_rest.rs

```diff
@@ -54,327 +68,370 @@ impl HttpConfig {
-            *response.status_mut() = hyper::StatusCode::FORBIDDEN;
+) -> Result<StatusCode, Error> {
+    if let Some((directory, file_name)) = separate_path_and_file_name(&paths.full) {
+        if let Err(err) = create_directories(directory) {
+            return Err(Error::Upload {
+                err: err.into(),
+                path: paths.request,
+            });
         }
 
-        let full_path = directories_path.join(file_name);
+        let full_path = directory.join(file_name);
 
         match stream_request_body_to_path(&full_path, request.body_mut()).await {
-            Ok(()) => {
-                *response.status_mut() = hyper::StatusCode::CREATED;
-            }
-            Err(_err) => {
-                *response.status_mut() = hyper::StatusCode::FORBIDDEN;
-            }
+            Ok(()) => Ok(StatusCode::CREATED),
+            Err(err) => Err(Error::Upload {
+                err,
+                path: paths.request,
+            }),
         }
     } else {
-        *response.status_mut() = hyper::StatusCode::FORBIDDEN;
+        Err(Error::InvalidPath {
+            path: paths.request,
+        })
     }
-    Ok(response)
 }
 
-async fn get(
-    request: Request<Body>,
-    file_transfer: &HttpConfig,
-) -> Result<Response<Body>, FileTransferError> {
-    let full_path = file_transfer.local_path_for_uri(request.uri().to_string())?;
-
-    if !full_path.exists() || full_path.is_dir() {
-        let mut response = Response::new(Body::empty());
-        *response.status_mut() = hyper::StatusCode::NOT_FOUND;
-        return Ok(response);
+async fn download_file(
+    paths: FileTransferPaths,
+) -> Result<StreamBody<ReaderStream<BufReader<File>>>, Error> {
+    let reader: Result<_, io::Error> = async {
+        let mut buf_reader = BufReader::new(File::open(paths.full).await?);
+        // Filling the buffer will ensure the file can actually be read from,
+        // which isn't true if it's a directory, but `File::open` alone won't
+        // catch that
+        buf_reader.fill_buf().await?;
+        Ok(buf_reader)
+    }
+    .await;
+
+    match reader {
+        Ok(reader) => Ok(StreamBody::new(ReaderStream::new(reader))),
+        Err(e) => {
+            if e.kind() == ErrorKind::NotFound || err_is_is_a_directory(&e) {
+                Err(Error::FileNotFound(paths.request))
+            } else {
+                Err(Error::FromIo(e))
+            }
+        }
     }
-
-    let contents = tokio::fs::read(full_path).await?;
-
-    Ok(Response::new(Body::from(contents)))
 }
 
-async fn delete(
-    request: Request<Body>,
-    file_transfer: &HttpConfig,
-) -> Result<Response<Body>, FileTransferError> {
-    let full_path = file_transfer.local_path_for_uri(request.uri().to_string())?;
-
-    let mut response = Response::new(Body::empty());
+// Not a typo, snake_case for: 'err is "is a directory"'
+fn err_is_is_a_directory(e: &std::io::Error) -> bool {
```

Comment:
```suggestion
fn err_kind_is_a_directory(e: &std::io::Error) -> bool {
```
Minor: If you wanna avoid that naming confusion

Reply:
That doesn't really avoid the issue, because we're still checking that the kind is "is a directory". Omitting the word kind makes it clearer that the input is a `std::io::Error` and not `std::io::ErrorKind`, and that is definitely a more ergonomically sensible choice here in my mind.